src/Biogeochemistry.jl

```diff
@@ -0,0 +1,138 @@
+import Oceananigans.Forcings: regularize_forcing, maybe_constant_field
+
+#####
+##### Generic fallbacks for biogeochemistry
+#####
+
+"""
+Update tendencies.
+
+Called at the end of calculate_tendencies!
+"""
+update_tendencies!(bgc, model) = nothing
+
+"""
+Update tracer tendencies.
+
+Called at the end of calculate_tendencies!
+"""
+update_biogeochemical_state!(bgc, model) = nothing
+
+@inline biogeochemical_drift_velocity(bgc, val_tracer_name) = nothing
+@inline biogeochemical_advection_scheme(bgc, val_tracer_name) = nothing
+@inline biogeochemical_auxiliary_fieilds(bgc) = NamedTuple()
+
+#####
+##### Default (discrete form) biogeochemical source
+#####
+
+abstract type AbstractBiogeochemistry end
+
+@inline function biogeochemistry_rhs(i, j, k, grid, bgc, val_tracer_name::Val{tracer_name}, clock, fields) where tracer_name
+    U_drift = biogeochemical_drift_velocity(bgc, val_tracer_name)
+    scheme = biogeochemical_advection_scheme(bgc, val_tracer_name)
+    src = biogeochemical_transition(i, j, k, grid, bgc, val_tracer_name, clock, fields)
+    c = @inbounds fields[tracer_name]
+        
+    return src - div_Uc(i, j, k, grid, scheme, U_drift, c)
+end
+
+@inline biogeochemical_transition(i, j, k, grid, bgc, val_tracer_name, clock, fields) =
+    bgc(i, j, k, grid, val_tracer_name, clock, fields)
+
+@inline (bgc::AbstractBiogeochemistry)(i, j, k, grid, val_tracer_name, clock, fields) = zero(grid)
+
+#####
+##### Continuous form biogeochemical source
+#####
+ 
+"""Return the biogeochemical forcing for `val_tracer_name` when model is called."""
+abstract type AbstractContinuousFormBiogeochemistry <: AbstractBiogeochemistry end
+
+@inline extract_biogeochemical_fields(i, j, k, grid, fields, names::NTuple{1}) =
+    @inbounds (fields[names[1]][i, j, k],)
+
+@inline extract_biogeochemical_fields(i, j, k, grid, fields, names::NTuple{2}) =
+    @inbounds (fields[names[1]][i, j, k],
+               fields[names[2]][i, j, k])
+
+@inline extract_biogeochemical_fields(i, j, k, grid, fields, names::NTuple{N}) where N =
+    @inbounds ntuple(n -> fields[names[n]][i, j, k], Val(N))
+
+@inline function biogeochemical_transition(i, j, k, grid, bgc::AbstractContinuousFormBiogeochemistry,
+                                           val_tracer_name, clock, fields)
+
+    names_to_extract = tuple(required_biogeochemical_tracers(bgc)...,
+                             required_biogeochemical_auxiliary_fields(bgc)...)
+
+    fields_ijk = extract_biogeochemical_fields(i, j, k, grid, fields, names_to_extract)
+
+    x = xnode(Center(), Center(), Center(), i, j, k, grid)
+    y = ynode(Center(), Center(), Center(), i, j, k, grid)
+    z = znode(Center(), Center(), Center(), i, j, k, grid)
+
+    return bgc(val_tracer_name, x, y, z, clock.time, fields_ijk...)
+end
+
+@inline (bgc::AbstractContinuousFormBiogeochemistry)(val_tracer_name, x, y, z, t, fields...) = zero(x)
+
+struct NoBiogeochemistry <: AbstractBiogeochemistry end
```

Comment:
why can't this not be a `Nothing`? (then in the model constructor you just pass a nothing)

Reply:
I think we probably don't need it since the only place a function is dispatched for it is:
```julia
required_biogeochemical_tracers(::NoBiogeochemistry) = ()
```
And I don't think there's a reason we cant just change that to be `bgc` or `::Nothing` like all of the other functions

Reply:
Found the reason for this, since we dispatch on the the model i.e. `(bgc::AbstractBiogeochemistry)(i, j, ...` we have to have some way to deal with there being no bgc.

The way we had this was to define `NoBiogeochemistry` so that it fell back to `@inline (bgc::AbstractBiogeochemistry)(i, j, k, grid, val_tracer_name, clock, fields) = zero(grid)`.

For consistency with how other optional features are handeled I could instread just add:
`@inline biogeochemistry_rhs(i, j, k, grid, ::Nothing, args...) = zero(grid)` which seems to solve the problem?

Reply:
This must also be more efficient than the origional `biogeochemsitry_rhs` for `NoBiogeochemistry`

Reply:
Just checked and this is how no particles are dealt with `update_particle_properties!(::Nothing, model, Δt) = nothing`